### PR TITLE
ARROW-10572: [Rust][DataFusion] Use aHash instead of FnvHashMap

### PR DESCRIPTION
**Pull request by Dandandan**
Use ahash https://github.com/tkaitchuck/aHash for hashing algorithm.

Difference is mainly visible in group_by benches:

``` 
aggregate_query_group_by 15 12                                                                             
                        time:   [2.0644 ms 2.0869 ms 2.1100 ms]
                        change: [-5.4843% -4.1892% -2.8525%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 2 outliers among 100 measurements (2.00%)
  2 (2.00%) high mild

Benchmarking aggregate_query_group_by_with_filter 15 12: Warming up for 3.0000 s
Warning: Unable to complete 100 samples in 5.0s. You may wish to increase target time to 9.6s, enable flat sampling, or reduce sample count to 50.
aggregate_query_group_by_with_filter 15 12                                                                             
                        time:   [1.8875 ms 1.9006 ms 1.9140 ms]
                        change: [-9.8182% -8.4717% -7.1023%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 5 outliers among 100 measurements (5.00%)
  1 (1.00%) high mild
  4 (4.00%) high severe
```

### CODE REVIEW

**Comment by alamb** [comment]
There appears to be a diff int he CI tests:
https://github.com/apache/arrow/pull/8654/checks?check_run_id=1392391221

```

---- execution::context::tests::count_distinct_integers_aggregated_single_partition stdout ----
thread 'execution::context::tests::count_distinct_integers_aggregated_single_partition' panicked at 'assertion failed: `(left == right)`
  left: `["a,3,2,2,2,2,2,2,2,2", "b,1,1,1,1,1,1,1,1,1", "c,3,2,2,2,2,2,2,2,2"]`,
 right: `["a,3,2,2,2,2,2,2,2,2", "c,3,2,2,2,2,2,2,2,2", "b,1,1,1,1,1,1,1,1,1"]`', datafusion/src/execution/context.rs:1077:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

---- execution::context::tests::count_distinct_integers_aggregated_multiple_partitions stdout ----
thread 'execution::context::tests::count_distinct_integers_aggregated_multiple_partitions' panicked at 'assertion failed: `(left == right)`
  left: `["b,5,4,4,4,4,4,4,4,4", "a,5,3,3,3,3,3,3,3,3", "c,1,1,1,1,1,1,1,1,1"]`,
 right: `["a,5,3,3,3,3,3,3,3,3", "c,1,1,1,1,1,1,1,1,1", "b,5,4,4,4,4,4,4,4,4"]`', datafusion/src/execution/context.rs:1105:9
```

Which is likely due to the change in order of the output (the values appear to be the same) which is not surprising given a change in the hash algorithm -- perhaps the test just needs to be updated

**Review by alamb — COMMENTED** [review]
Thanks for the contribution @Dandandan

This seems like a reasonable change to me. I didn't verify the benchmarks but I checked out  https://crates.io/crates/ahash and it appears legit though less widely used than `fnv`

We need to fix the CI tests, but otherwise I would ok with merging this in. What do others think?

**Comment by vertexclique** [comment]
I prefer using t1ha than ahash, which proven to be sound.

**Comment by paddyhoran** [comment]
> I prefer using t1ha than ahash, which proven to be sound.

The `t1ha` crate is `Licensed under zlib License`.  I don't think that is compatible with Apache (`ahash` is).

**Comment by andygrove** [comment]
Do we have a feel for the performance implications of this change for large data sets as opposed to the micro benchmarks?

**Comment by Dandandan** [comment]
A nice overview is listed here: https://github.com/tkaitchuck/aHash/blob/master/compare/readme.md comparing aHash to other algorithms. aHash passes the full suit of https://github.com/rurban/smhasher , FnvHash has some weak points.

 @andygrove any way to test it? I recently also used it for the polars project where it brought around 20-30% speed ups to group by and hash join vs SeaHasher: https://github.com/ritchie46/polars/pull/128

**Comment by andygrove** [comment]
@Dandandan Thanks for the links. That addresses my concern. We do have a benchmark crate in this repo with instructions for running a TPC-H with larger data sets but I don't see any reason not to go ahead and merge this.

**Review by jorgecarleitao — COMMENTED** [review]
Thanks a lot @dandandan for this! 🚀 

LGTM! The tests fail for obvious reasons (no sort in place), which is expected with this change. I do not think that we need to enforce that an aggregation is sorted, and we have a `sort` for users to apply in case they want them sorted, so, all good.

Tests need fixing with the new behavior (un-sorted aggregate results). I think that we could sort the testing results in native rust (as opposed to in datafusion), so that when a test fails, it is easier to figure out whether it was the `sort` node or the `aggregate` node that caused the issue.

**Comment by alamb** [comment]
> Tests need fixing with the new behavior (un-sorted aggregate results). I think that we could sort the testing results in native rust (as opposed to in datafusion), so that when a test fails, it is easier to figure out whether it was the sort node or the aggregate node that caused the issue.

In past lives, we have addressed the "output order of group by may not be sorted" by normalization in the test suite itself (aka sort the output record batch in the *test* itself).

**Comment by andygrove** [comment]
I ran the TPC benchmark and saw no noticeable difference in performance, so LGTM.

**Comment by Dandandan** [comment]
> I ran the TPC benchmark and saw no noticeable difference in performance, so LGTM.

Thanks. At least no noticeable regression then.

File: rust/datafusion/src/physical_plan/hash_aggregate.rs [hunk 1]
```diff
@@ -882,25 +867,12 @@ mod tests {
...
+        rows.sort();
+
         assert_eq!(
-            *b,
-            Float64Array::from(vec![
-                1.0,
-                (2.0 + 3.0 + 2.0) / 3.0,
-                (3.0 + 4.0 + 4.0) / 3.0
-            ])
+            rows,
```
Comment (jorgecarleitao) [inline rust/datafusion/src/physical_plan/hash_aggregate.rs:874]: Could we keep the expressions? `(2.0 + 3.0 + 2.0) / 3.0` is much more informative than `2.3333333333333335`, imo.

Comment (alamb) [reply rust/datafusion/src/physical_plan/hash_aggregate.rs:874]: The tests changed to be string comparisons so I am not sure we can keep the arithmetic anymore. 

Maybe we could put it in comments, something like

```
            vec![
                    "2,1.0", 
                    "3,2.3333333333333335",  // 3, (2 + 3 + 2) / 3
                    "4,3.6666666666666665" // 4, (3 + 4 + 4) / 3
            ]
```

**Review by alamb — APPROVED** [review]
I think it looks good to me

File: rust/datafusion/src/physical_plan/hash_aggregate.rs (hunk 1 above)
Comment (Dandandan) [reply rust/datafusion/src/physical_plan/hash_aggregate.rs:874]: Makes sense, added the comments

**Comment by alamb** [comment]
I think this is good to go -- what do you think @jorgecarleitao  / @andygrove ?

**Review by jorgecarleitao — APPROVED** [review]
LGTM!

**Pull request closed without merging**